src/eckit/types/Time.cc

```diff
@@ -27,80 +30,98 @@ inline void printTime(std::ostream& s, long n) {
-            // hh:mm:ss
-
-            err = result[0].length() != 2 || result[1].length() != 2 || result[2].length() != 2;
-
-            hh = atol(result[0].c_str());
-            mm = atol(result[1].c_str());
-            ss = atol(result[2].c_str());
-
-            break;
-
-        default:
-            err = true;
-            break;
+        }
     }
-
-    if (err) {
-        throw BadTime(std::string("Invalid time ") + s);
+    else {
+        if (std::regex_match (s, m, std::regex("^-?[0-9]*\\.[0-9]+$"))) { // floating point (hours)
+            long sec = std::round(std::stod(s)*3600);
+            hh = sec/3600;
+            sec -= hh*3600;
+            mm = sec/60;
+            sec -= mm*60;
+            ss = sec;
+        }
+        else {
+            if (std::regex_match (s, m, std::regex("^([0-9]+):([0-5]?[0-9])(:[0-5]?[0-9])?$"))) {
+                for (int i=1; i<m.size(); i++) {
+                    if (m[i].matched) {
+                        switch (i) {
+                            case 1: hh = std::stol(m[i].str()); break;
+                            case 2: mm = std::stol(m[i].str()); break;
+                            case 3: std::string aux = m[i].str();
+                                    aux.erase(0,1);
+                                    ss = std::stol(aux); break;
+                        }
+                    }
+                }
+            }
+            else {
+                if (std::regex_match (s, m, std::regex("^-?([0-9]+[dD])?([0-9]+[hH])?([0-9]+[mM])?([0-9]+[sS])?$"))) {
+                    for (int i=1; i<m.size(); i++) {
+                        if (m[i].matched) {
+                            std::string aux = m[i].str();
+                            aux.pop_back();
+                            long t = std::stol(aux);
+                            switch (i) {
+                                case 1: dd = t; break;
+                                case 2: hh = t; break;
+                                case 3: mm = t; break;
+                                case 4: ss = t;
+                            }
+                        }
+                    }
+                    ss += 60 * (mm + 60 * (hh + 24 * dd));
+                    if (s[0] == '-') {
+                        ss = -ss;
+                    }
+                    dd =  ss / 86400;
+                    hh = (ss /  3600) % 24;
+                    mm = (ss /    60) % 60;
+                    ss =  ss          % 60;
+                } else {
+                    std::string msg = "Wrong input for time: ";
+                    msg += s;
+                    throw BadTime(msg);
+                }
+            }
+        }
     }
 
-    if (hh >= 24 || mm >= 60 || ss >= 60 || hh < 0 || mm < 0 || ss < 0) {
+    if (mm >= 60 || ss >= 60 || (!extended && (hh >= 24 || dd > 0 || hh < 0 || mm < 0 || ss < 0))) {
         std::string msg = "Wrong input for time: ";
         Translator<long, std::string> t;
+        if (dd>0) {
```

Comment:
This isn't consistent with the check above, for ss<86400. (eg, line 35 but there are more). If the class isn't to support days, and it also doesn't have and accessor for it, this should actually fail (or, remove the checks and support it)

Reply:
Also, this printing of time as string is repeated in multiple places -- ideally, these places should be replaced with a operator std::string (that you already have)